### ISSUE TITLE
Center OK Button Issues

### ISSUE THREAD

**Issue by theancientways** [sspanak/tt9#393]
Hey, I'm encountering an issue on the Sonim XP3plus where the center ok button doesn't work sometimes in the contacts app. The only time it consistently doesn't work is when I try to use the "add a contact" function from the text message app. It takes me to a window with a search field above a list of contacts, the top option of which being "create a contact". This is the window where the center button consistently doesn't work. It does work if I switch to the phone's native text input method. I know you didn't test this on every phone so there's bound to be bugs here an there on different devices but still thought I'd report it.

**Comment by theancientways** [comment]
Some more info... Oddly, after toggling off the "send messages with OK in messenger", the center button seems to function in the contacts app only once, after that it doesn't work. Weird.

**Comment by sspanak** [comment]
I don't quite get what's wrong, because you just say "it doesn't work", but I sounds really similar to [this bug](https://github.com/sspanak/tt9/pull/364). If it is the same, it will be fixed in the next version. Or is it different?

**Comment by theancientways** [comment]
Hey, I don't know if I have the know-how to be able to explain it beyond how I did in my first post, basically in a nutshell the center OK button on my flip phone doesn't work when a new number texts me and I try to add them as a new contact. From the text message thread I press a soft key and select "add new contact" from the popup list. That takes me to the contacts app with a button that says "add new contact", trying to press this with the OK button does nothing. Sorry I can't explain better but I could get screenshots if that might help. It's not the worst of inconveniences but it's just odd that it doesn't work right.

If I download your new release, will it carry over my user dictionary or will I have to start it from scratch?

Thanks as always.

**Comment by theancientways** [comment]
hey just a bump for my question at the end of my above post, is there a way to save your user dictionary when you upgrade or do you have to start from scratch?

**Comment by sspanak** [comment]
This is quite problematic at the moment. The upcoming version 27.0 will introduce major database changes, which will make it possible to protect the added words. It would even be possible to delete all languages, then reload them, but the words will remain safe. In the future, it would be possible to export them, as well.

However, the above comes at the cost of deleting absolutely all words when upgrading to v27.0.

But still, you may actually win more than you lose after the upgrade, depending on the language(s) you are using. Bulgarian, English, Romanian, Russian and Ukrainian will receive a major update including thousands of new words.

I realize many other languages lack rich dictionaries, but hopefully, with community contributions, it will get better over time.

**Comment by theancientways** [comment]
I'm willing to start from scratch to get access to those new features. Sounds like v27 will be a huge leap. Looking forward to it and thanks for all your work! Will donate.

**Comment by sspanak** [comment]
As I mentioned in the other issue, the new version is out. Please check it out and let me know if it fixes the problem.

**Comment by drol845** [comment]
I am using tt9 on a **Sonim XP3800**. The center button isn't working correctly in some apps. For example, in the calculator, the center button represents "equals", but it doesn't return the calculation result. In the music player, long-pressing the center button over a song doesn't open the menu. This issue occurs in some other apps as well.

Overall, I'm very excited with tt9, I'm only switching back to the original keyboard when I need a calculator, etc.

It would be great if there would be an easier way to switch keyboards back and forth.

**Comment by sspanak** [comment]
> It would be great if there would be an easier way to switch keyboards back and forth.

There is an easy way to switch from TT9 to another keyboard. See [the manual](https://github.com/sspanak/tt9/blob/master/docs/user-manual.md#change-keyboard-key-default-unassigned). However, I am not sure if the Sonim native keyboard allows the same as easily.

> I am using tt9 on a Sonim XP3800. The center button isn't working correctly in some apps. For example, in the calculator, the center button represents "equals", but it doesn't return the calculation result. In the music player, long-pressing the center button over a song doesn't open the menu. This issue occurs in some other apps as well.

I remember another problem with the calculator in [this PR](https://github.com/sspanak/tt9/pull/326/files), but there seem to be many more problems on Sonim phones, than I thought. I've created a special file for application and device ~~hacks~~ compatibility options, so I can easily add more, but it's incredibly difficult to diagnose and fix them remotely.

Have you tried any other 3rd party keyboard, for example [k9t9](https://github.com/danielshorten/k9t9)? If not and you are willing to help with debugging, please check it out. If the player and the calculator work properly, then there is something wrong with Traditional T9. I can then try to figure out how k9t9 handles OK and do the same. If it doesn't work either, it means the Sonim apps and keyboard are meant to work together and send some special signals to each other for improved usability. In this case, I'll have to guess what is going on and apply some hack.

As you probably understand, I can't promise much. I think I have an idea about a couple of experimental fixes, that I intend to release in the upcoming v29.0, so, again, if you are willing to help, follow this issue. I'll make an announcement and let you know how to enable them. Then you'll just have to tell me which one works, if it works.

**Comment by theancientways** [comment]
> I am using tt9 on a **Sonim XP3800**. The center button isn't working correctly in some apps. For example, in the calculator, the center button represents "equals", but it doesn't return the calculation result. In the music player, long-pressing the center button over a song doesn't open the menu. This issue occurs in some other apps as well.
> 
> Overall, I'm very excited with tt9, I'm only switching back to the original keyboard when I need a calculator, etc.
> 
> It would be great if there would be an easier way to switch keyboards back and forth.

I'm experiencing the same issues on the Sonim XP3plus. It's weird in the calculator, when you press enter, the "equals" function is executed when you press the "ok" button, it just displays a blank space. You can tell the equals function is executed because it shifts the equation up as it should when you press the ok button. As I stated in my original post, when in TT9 I can't add a phone number as a new contact from the text message app, it takes me to the "add a contact" screen, but I can't execute the function as pressing the "ok" button does nothing. To me it's not a huge deal, I just go into the TT9 app settings and switch to the native keypad then go back and execute the function I needed. This is a compromise I can live with.

**Comment by asherGe** [comment]
> Have you tried any other 3rd party keyboard, for example [k9t9](https://github.com/danielshorten/k9t9)?

@sspanak, I am experiencing the same issues that @drol845 and @theancientways  are describing on the XP3plus (XP3900) with the calculator app and the contacts app (tt9 v28.0). I took your suggestion and installed k9t9. While k9t9 appears to have other issues with this phone (e.g. backspace/delete not working), it DOES allow the OK button to work properly both for the calculator and contacts app. So, it appears there IS potential that you can solve this issue?!

> There is an easy way to switch from TT9 to another keyboard. See [the manual](https://github.com/sspanak/tt9/blob/master/docs/user-manual.md#change-keyboard-key-default-unassigned). However, I am not sure if the Sonim native keyboard allows the same as easily.

The native keyboard does not give option for hotkey to access keyboard selection. You have made it easy to switch to another input method with the hotkey shortcut but must go through the Settings->System-Languages & Input -> On-screen keyboard -> Manage on-screen keyboards -> [toggle on / off ]  and then also go into tt9 from app menu -> Initial Setup - Select Default Keyboard to re-select tt9 as default to switch back to tt9.

**Comment by asherGe** [comment]
> I don't quite get what's wrong, because you just say "it doesn't work", but I sounds really similar to https://github.com/sspanak/tt9/pull/364. If it is the same, it will be fixed in the next version. Or is it different?

Here's my take at describing the Contacts problem: Say someone who I don't have in my contacts calls me and I want to add them to contacts. I can go to Phone -> Recent Calls -> select phone number of recent call. Then there is a soft-key option "Add to Contacts". This takes me to a screen with the alphabetical list of my contacts. At the top of the screen is a search field and I can type in that field to filter the contacts listed below and the OK button works correctly inside the search field. However, in this scenario, when the search field is empty, the first option below it is "Create new contact", followed by the alphabetical list of contacts. The text cursor is in the search field and the highlighted down-selection is "Create new contact" but the OK button does not work - does nothing - and it is not possible to execute the "create new contact" function. Furthermore, if I want to add the number to an existing contact, I would use the search field to filter the selection list and when I navigate using up/down keys to select desired contact, the OK button does not execute the function to add the number to that contact - the okay press does nothing.   Appears that the issue has something to do with the active search field where the OK button is trying to perform while one is trying to use it to select a down-list selection beneath the search field.

**Comment by sspanak** [comment]
> While k9t9 appears to have other issues with this phone (e.g. backspace/delete not working), it DOES allow the OK button to work properly both for the calculator and contacts app. So, it appears there IS potential that you can solve this issue?!

Yes, I think I know what is wrong. I just have to be careful not to cause the same on TCL Flip 2 (#182), or any of: #210, #309 (similar problem on another Sonim), or: #77, #51 (OK must submit forms properly, navigate to web pages in browsers and whatnot).

**Notes to self:** when any of DPAD_CENTER, ENTER or NUMPAD_ENTER is pressed, and the suggestions are hidden, and no app hack is applied, then return false.

**Note to everyone:** I can't provide a fix immediately, because the code is very unstable now. I'll have to complete v29.0 first. I've been thinking that v30.0 could be dedicated to fixing such device-specific problems anyway.

**Comment by theancientways** [comment]
> I am using tt9 on a **Sonim XP3800**. The center button isn't working correctly in some apps. For example, in the calculator, the center button represents "equals", but it doesn't return the calculation result. In the music player, long-pressing the center button over a song doesn't open the menu. This issue occurs in some other apps as well.
> 
> Overall, I'm very excited with tt9, I'm only switching back to the original keyboard when I need a calculator, etc.
> 
> It would be great if there would be an easier way to switch keyboards back and forth.

Hey I just found out if you go to the advanced panel the calculator displays the answer, then if you switch back to the basic panel, the answer gets displayed but not consistently. It's only consistently displayed in the advanced panel, so that's a way to still be able to use the calculator in on the XP3plus with TT9 enabled.

**Comment by sspanak** [comment]
@drol845 , @theancientways , @asherGe , please try out [v29.15](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v29.15.apk). I have attempted to fix the problem there. Any feedback will be much appreciated.

**Comment by asherGe** [comment]
Hi @sspanak,

Thanks so much for attempting this fix. I have v29.15 loaded on the Sonim
XP3 plus (xp3900). The issues persist with both the Calculator app and the
Contacts app.

In the Calculator app, when you press okay (=) it appears that an extra
line is being inserted and the answer to the calculation is scrolling up
off the visible screen.

In the Contacts app, when trying to add new contact, the OK button works in
the search field for tt9 text operations but does not work as an enter key
to add entered text as a new contact nor does it allow to select an
existing contact from drop down list of options matching the search. In the
second case where there are multiple options of existing contacts, the
first entry in the drop down list is highlighted initially as the
selection, up down arrow keys work to change selection, but okay key does
not enter the selection as the selected contact to add the phone number to.

Keep me posted on anything more I can do to help troubleshoot.

Thanks,

Asher
---------------------------------------------------------------
Asher Gelbart
Green Energy Now
***@***.***
575.574.7119
---------------------------------------------------------------

**Comment by sspanak** [comment]
No, not at this time. I thought I had it figured out, but I am still missing something. I'll give it a try later.

**Comment by asherGe** [comment]
Hi @sspanak/tt9
***@***.***>

One other piece of info... the v29.15 mod also broke the display of entry
mode and suggestions at bottom of screen upon first push of okay button in
messaging text field. Instead of displaying the entry mode and suggestions,
it returned an extra line in the text field. I just reverted back to v29.0.

Thanks,

Asher

---------------------------------------------------------------
Asher Gelbart
Green Energy Now
***@***.***
575.574.7119
---------------------------------------------------------------

**Comment by sspanak** [comment]
> One other piece of info... the v29.15 mod also broke the display of entry
mode and suggestions at bottom of screen upon first push of okay button in
messaging text field. Instead of displaying the entry mode and suggestions,
it returned an extra line in the text field.

I've noticed this as well, but it happens only in Duolingo for me. I can test there and fix it.

But what happens if you just start typing? Won't the UI appear? In v29.0 I've tried to reduce the number of applications where pressing OK is mandatory. In most of them, it should be possible to make TT9 appear with any key from 0 to 9.

**Comment by asherGe** [comment]
Hi @sspanak,

Unfortunately no, (in v29.0) in the messaging app the UI only appears if I
press OK when the text field is empty (says "Type message" in grey
letters). If I start typing the UI does not appear, but if I backspace to
delete the letters, once I've emptied the text field "Type message"
reappears, and then if I press OK, the UI shows up.
---------------------------------------------------------------
Asher Gelbart
Green Energy Now
***@***.***
575.574.7119
---------------------------------------------------------------

**Comment by asherGe** [comment]
Hi @sspanak,

I noticed that in some of the native apps on the XP3plus the center top
soft key (between screen and up arrow key) is sometimes used for OK entry.
Here is an example (see attached screen shot). This is in the clock - alarm
app. The app uses left / right and up / down arrow keys to adjust the alarm
set time, the center OK button has no response, and OK is shown to use the
center soft key. Anyway, I understand that you are trying to make the
adjustments to the code to work universally across all phones and all apps,
so perhaps this is just extraneous info. Wondering if it may be possible
and helpful to assign an alternate key (such as the center soft key) the OK
operation when the normal OK press does not operate as intended?

---------------------------------------------------------------
Asher Gelbart
Green Energy Now
***@***.***
575.574.7119
---------------------------------------------------------------

**Comment by drol845** [comment]
@asherGe The older model XP3 dosn't have these soft keys.

**Comment by sspanak** [comment]
> Here is an example (see attached screen shot)

Something has gone wrong with the attachments, I see none of them. Could you please try posting them again?

> Anyway, I understand that you are trying to make the
adjustments to the code to work universally across all phones and all apps,
so perhaps this is just extraneous info.

Perhaps, perhaps not. It may help me understand better how the Sonim phone works.

> Wondering if it may be possible
and helpful to assign an alternate key (such as the center soft key) the OK
operation when the normal OK press does not operate as intended?

I am thinking of this. But first, I would like to review and try to fix the OK key one more time. I was able to reproduce the incorrect behavior in Duolingo and the Qin messaging app. I may have made some incorrect assumptions about printing a new line vs performing some action.

**Comment by asherGe** [comment]
Apologies, I was replying directly from e-mail which appears not as good as commenting from github. Here's the attachment
![Clock_Alarm_ScreenShot](https://github.com/sspanak/tt9/assets/159474592/1f245417-a86a-4bcf-abce-bfd9ea6c2d3f)

**Comment by sspanak** [comment]
I see. This is similar to the Kyocera Digno problem, #423. This may or may not be fixed, I am not sure.

The good new is, thanks to your donations, I was able to buy an XP3800 and I believe I fixed all problems with the OK key! I didn't get what is wrong with the calculator, though. I works just fine for me. If you still think it isn't, would it be possible to record a video, so I can see it?

Please, try out the new [v30.0](https://github.com/sspanak/tt9/releases/tag/v30.0) and let me know if anything is still broken. And have in mind, I have the non-plus version, meaning, support for the soft keys is questionable at this time.

**Comment by asherGe** [comment]
Excellent! Thanks. The contacts and calculator issues with OK discussed above are now resolved with XP3900.

However, I can no longer get the suggestion list or input mode to appear at bottom of screen. Looks the same when I choose stealth mode or suggestion list only. In standard text messaging app, the tt9 input icon appears immediately when I select a text thread, but then I can't activate the UI with the press of the okay button as I was able to in all previous versions. 

The tt9 input icon appears when I have text fields in Messaging and Contacts apps but not in the calculator or clock apps with number fields  (is it supposed to?).

**Comment by sspanak** [comment]
Well, Stealth mode is... stealth, you will not see them. I guess, I should rename it to "Invisible".

> the tt9 input icon appears immediately when I select a text thread, but then I can't activate the UI with the press of the okay button as I was able to in all previous versions.

You don't need to. If the icon has appeared, you can start typing right away and the UI will appear automatically. Activation using OK was a backup plan because sometimes the UI wouldn't show up at all. But I see what you mean. I think I can make a small improvement to recreate the old experience (well, as much as possible).

> The tt9 input icon appears when I have text fields in Messaging and Contacts apps but not in the calculator or clock apps with number fields (is it supposed to?).

Yes, this is correct. The icon indicates when TT9 is in charge of typing. In some apps, such as dialers, calculators, and whatnot, it is best to let _them_ type, because they have sort of a simple keyboard built inside of them. If the main Android keyboard also tries to produce text or numbers at the same time, it leads to double letters (or numbers). There is even a bug for this: #300. There is a short video attached in one of the last posts, where you can see double numbers appearing per a single key press.

By the way, there is no need to turn on the status icon on Sonim phones. It is there for Qin F21 Pro, which prohibits holding keys when there is no icon. As you can imagine, this effectively disables some of TT9's functions. But anyway, if you just like to know when an app has requested user input, and TT9 has acknowledged it can work together with the app, keep the icon, there is no harm.

**Comment by asherGe** [comment]
I should clarify, the UI does not appear when I start typing…it acts like stealth mode in whichever mode I have selected. That is great if OK no longer needed to wake it up, but it isn’t appearing with initiation of typing.

**Comment by asherGe** [comment]
In previous versions with this phone, UI would not initiate with typing. The only way to get the UI to display was to press OK while text field was empty. If I started typing, no UI, but if I backspaced to clear the text field and then pressed OK, it would appear and remain visible for the remainder of the text entry. In current version, I can’t get the UI to display at all.

**Comment by sspanak** [comment]
I am not sure why isn't it working on your phone. The UI appears both on my XP3 and my other Android 10 and 11 phones.

Restoring the old behavior is not a good idea, but I've come up with an different way of showing the UI. If you are willing to install a beta version, here is [v30.10](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.10.apk) for you to test. After installing it, you will be able to show the UI like this:

1. Make absolutely sure, you have selected any of the 3 visible modes from Settings -> Appearance.
2. Start typing. Even one letter is enough.
3. If it stays hidden, press OK, while you are still composing a new word (while it is underlined) and the suggestion list should appear.

If it still doesn't want to, probably the standard procedure for Android 9+ is not the way to go on the XP3plus. I'll review the old code again and come up with an alternative solution.

**Comment by asherGe** [comment]
Hi sspanak, again, thanks for your efforts here! Unfortunately this attempt (v30.10) did not work. Still can't get UI to appear. I can see the difference in behavior with OK between the three visible modes and the invisible mode, but UI won't appear. In the three visible modes, when I press OK while word is underlined, nothing happens and the word remains underlined. The 0 (space) key selects the word. In the invisible mode, the OK does the same operation as the space (0) key.

**Comment by sspanak** [comment]
> Unfortunately this attempt (v30.10) did not work. Still can't get UI to appear.

Alright, so far we have tried two ways of showing the main UI, two more remain. Here is an APK with the third possible way: [v30.14](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.14.apk). You can use the same technique: start typing, then press OK and this time it should appear. The fourth option seems really unlikely, so it must be this one.

I've tried enabling the show function when there is no text and one simply presses OK, but this causes unnecessary flash of the UI in applications with search fields, so I gave up the idea. I think pressing OK after starting to type is good enough.

>  thanks for your efforts here!

I must thank you for the patience and for being my right hand in solving this last problem. :slightly_smiling_face:

**Comment by asherGe** [comment]
Okay, getting close! So the UI DOES appear now and stays active on the first press of OK while having typed at least one letter in the text field, but it also sends whatever has been typed as a completed message (which I think is unintended and a bit annoying?)...after that, once the UI has been activated, OK works as intended to select/approve underlined word and only sends the message when nothing is selected (i.e. on a second consecutive press of OK).

**Comment by sspanak** [comment]
Alright, here is what I can do. I cannot prevent sending if there is text. However, I can make OK display the UI if there is no text (like before), but in that case, it will not send blank text. I wanted to keep that functionality for some apps that display all results when you search for nothing, but I guess this is not so important, neither really intuitive.

**Comment by sspanak** [comment]
As promised, here is a new APK: [v30.15](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.15.apk). It allows showing the UI by pressing OK when there is no text and also some more small tweaks. Showing after starting to type is no longer possible to avoid accidental sending. Overall, it should be much more intuitive, I've tested both on Sonim XP3800 and on Qin F21 Pro+ and it is fine, "it just works". I really hope it is OK on the XP3900 too.

**Comment by asherGe** [comment]
Yup! Excellent. It works on the XP3900 too!

**Comment by asherGe** [comment]
v30.15 working pretty much flawlessly now on XP3900 in the native messaging app. There is one observation on the ADD Word (*-key) feature recently mentioned in the #495 thread. It now works when UI is activated but *-key-press is unresponsive when the UI is invisible (in one of the three visible modes prior to activation with OK press and in invisible mode). 

So, one more OK button issue cropped up in the Contacts app with this latest version v30.15.  Curious if it replicates on the XP3800. If you search for a contact by typing in the search field, the downlist of filtered selections appears but the OK button no longer works to select the contact. If you leave the search field empty, you can scroll through all the unfiltered complete list of contacts and OK still works. I have well over 500 contacts on my phone so the ability to use the filter is essential to quickly finding a contact phone number.  

Currently, the contacts app allows the UI to appear in certain circumstances but not in others. Pressing OK in the empty search field doesn't activate the UI, instead it selects the first contact in the down list.  If I choose to ADD a new contact and go to the first Name field and press OK, the UI activates.  The incompatibility issues appear to be occurring where there is an ambiguity between OK selecting from a downlist or making a tt9 OK entry in the search text field (empty or filtered).

**Comment by wazroth** [comment]
Works as described for me (XP3900) in the messenger compose window and notes app. The interface shows by default in the URL field of the browser.

Small 3900 issues, not sure whether addressable: 
- In the compose interface for the messenger app, in the "To" field, pressing OK pulls up the UI _and_ moves focus to the "Message" field.
- In the Contacts app: context: the cursor is always present in the "Find contacts" field but a contact is also always highlighted in results window; navigating with up/down changes the selection in the results. The bug is that OK always selects the currently highlighted contact, and it's impossible to pull up the menu.
- In the Phone app, pressing OK in the empty Search field moves focus from the Search field to the recent calls browser, and does not bring up the interface.

**Comment by neutralinsomniac** [comment]
Just to add on to the fun, using v30.15, the OK button doesn't work in Signal's message input box at all. To select the current active completion, I have to type space and then back (delete). I'm not sure if Signal is one of those apps that eats the OK button for its own purposes (it *seems* to do nothing ever) or if this is something that tt9 should be able to handle.

**Comment by neutralinsomniac** [comment]
> v30.15 working pretty much flawlessly now on XP3900 in the native messaging app. There is one observation on the ADD Word (*-key) feature recently mentioned in the #495 thread. It now works when UI is activated but *-key-press is unresponsive when the UI is invisible (in one of the three visible modes prior to activation with OK press and in invisible mode).
> 
> So, one more OK button issue cropped up in the Contacts app with this latest version v30.15. Curious if it replicates on the XP3800. If you search for a contact by typing in the search field, the downlist of filtered selections appears but the OK button no longer works to select the contact. If you leave the search field empty, you can scroll through all the unfiltered complete list of contacts and OK still works. I have well over 500 contacts on my phone so the ability to use the filter is essential to quickly finding a contact phone number.
> 
> Currently, the contacts app allows the UI to appear in certain circumstances but not in others. Pressing OK in the empty search field doesn't activate the UI, instead it selects the first contact in the down list. If I choose to ADD a new contact and go to the first Name field and press OK, the UI activates. The incompatibility issues appear to be occurring where there is an ambiguity between OK selecting from a downlist or making a tt9 OK entry in the search text field (empty or filtered).

I have the same OK button not working in the contacts list after filtering issue on the XP3900.

**Comment by neutralinsomniac** [comment]
For both of the above issues (OK button not working in Signal, OK button not working in Contacts), both apps work great in v30.0. Obviously the on-screen completions don't happen on that version, but otherwise everything seems great.

**Comment by sspanak** [comment]
OK, here is an APK ([v30.16](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.16.apk)), where you can select between the possible ways of displaying the UI. Anyone willing to keep testing, go to Settings, scroll down to the version info and click it to open the debugging settings. In there, look for "Show Window Mode". Try out the different modes and let me know of the results.

Some considerations:
1. The modes that do DPAD_CENTER emulation may send messages or perform some other action. Test them in a safe application or disable the SIM card.
2. The "auto-select" mode ought to work on all phones and it is the one I want to keep. Please try that one first.

> There is one observation on the ADD Word (*-key) feature recently mentioned in the https://github.com/sspanak/tt9/issues/495 thread. It now works when UI is activated but *-key-press is unresponsive when the UI is invisible (in one of the three visible modes prior to activation with OK press and in invisible mode).

This is by design. Some keys are deliberately disabled when TT9 is hidden.

> So, one more OK button issue cropped up in the Contacts app with this latest version v30.15. Curious if it replicates on the XP3800

Searching and selecting works fine. However, if I click a contact, then go back to the list, I can only type numbers. To be able to search normally again, I have to close the app and reopen it. This is not critical and is not a problem with the OK key, but with the entire application, so I'll check it out later.

@wazroth, @neutralinsomniac, I hope at least one of the Show Window modes will work properly and solve all the small problems. The Contacts app is an exception, but it should work fine most of the time, as I explained.

**Comment by neutralinsomniac** [comment]
Alright here's the results for v30.16:

Messaging app:

| Window Mode | Suggestions window shows on OK | Completions can be selected with OK | Message can send on OK |
| --- | --- | --- | --- |
| Auto select | :x: | :white_check_mark: | :white_check_mark: |
| Android 8 way | :x: | :white_check_mark: | :white_check_mark: |
| Android 9 way | :x: | :x: | :x: |
| DPAD_CENTER emulation | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| Android 8 way + DPAD_CENTER emulation | :x: | :white_check_mark: | :white_check_mark: |

Signal:

| Window Mode | Suggestions window shows on OK | Completions can be selected with OK | Message can send on OK |
| --- | --- | --- | --- |
| Auto select | :x: (1) | :white_check_mark: | :white_check_mark: |
| Android 8 way | :x: | :white_check_mark: | :white_check_mark: |
| Android 9 way | :x: | :x: | :x: |
| DPAD_CENTER emulation | :x: (2) | :white_check_mark: | :white_check_mark: |
| Android 8 way + DPAD_CENTER emulation | :x: (1) | :white_check_mark: | :white_check_mark: |

1: On initial OK, nothing happens, but on second and subsequent OK's, Signal says "Message is empty!" which suggests that tt9 *tried* to display itself but just couldn't for some reason. As a note, the stock 12-key keyboard is able to display suggestions while typing in Signal.
2: On initial OK *and* subsequent OKs, nothing happens.

I am curious now that I'm thinking about this, the stock 12-key keyboard displays its own suggestions window whenever you start typing (not needing to "wake up" the keyboard with a specific key like OK). Is that keyboard doing something different from what tt9 is attempting?

**Comment by asherGe** [comment]
neutralinsomniac - great matrix! Too bad still compatibility issues with Signal (I don't use it myself). 

DPAD_CENTER emulation so far appears to solve all the issues that I have experienced in Messaging App, Contacts App, and Calculator App (with XP3900). Everything works as expected and I can even get the UI to display when pressing OK from text entry field (other than search field) in the Contacts App and the Calendar App. This mode offers the best experience so far with the XP3900. 

sspanak - you've mentioned for future releases the possibility of adding another option in the Compatibility options in Initial Setup. Adding this DPAD_CENTER emulation option there would allow tt9 to be maximally compatible with the XP3900 and still allow you to maintain "auto select" as the default.

**Comment by sspanak** [comment]
Thank you very much for the cooperation!

Based on your feedback and more digging in the Android documentation, I've come up with two more ways of displaying the UI. Please, try [v30.17](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.17.apk). You can find two more "Show Window" options: 
- "Use CandidateView". I think this is what the native keyboard does. It feels a bit sluggish and it occasionally produces a 1px white bar under the suggestions list, but it works fine on the XP3800, including also solving all problems with the Contacts app. Also, it works reliably on the Qin F21 Pro.
- "Android 9 (no flags)". According to the documentation this is the way to put as much pressure on Android to display the active keyboard. However, it **does not** work for me. You may want to try it out, just in case.

Note that, when switching to and from "CandidateView" mode, you may need to go to the home screen then reopen the Messages or Notes application, otherwise it may not switch correctly. 

> sspanak - you've mentioned for future releases the possibility of adding another option in the Compatibility options in Initial Setup. Adding this DPAD_CENTER emulation option there would allow tt9 to be maximally compatible with the XP3900 and still allow you to maintain "auto select" as the default.

I don't like the idea of emulating a key press. As noted above, depending on the application, it may have a side effect, like switching focus to something else or even attempting to send a message. I'll include it only if there is no other choice.

> I am curious now that I'm thinking about this, the stock 12-key keyboard displays its own suggestions window whenever you start typing (not needing to "wake up" the keyboard with a specific key like OK). Is that keyboard doing something different from what tt9 is attempting?

Here are some technical details for your curiosity then.

If you can type, TT9 is awake, there is no real waking up. "Waking up" is just a more human-friendly expression for "displaying the main user interface". 

The problem is I am not doing much but calling a function named "showWindow()" or "requestShowSelf()", depending on the Android version. Each of these is supposed to draw the keyboard on the screen and they work perfectly fine on all the devices I have, but apparently Sonim have modified the Android code in their newer phones and now none of the functions work. They may have introduced a third one, I don't know.

I've noticed some applications respond to the DPAD Center key and send requests to Android to show the keyboard. And the native applications know exactly what to do, so this works in Messages and Contacts, but it does not in Signal.

**Comment by neutralinsomniac** [comment]
Alright so, Android 9 way (no flags), inset size 0, completions come up immediately when I start typing (no need to hit OK first) in both Signal and the stock messaging app, except I then can't select completions with OK. UseCandidatesView still doesn't show completions in any app. The strings to match this phone are "Sonimtech" "XP3900".

**Comment by asherGe** [comment]
Agree, Can't get UI to appear in CandidateView with any of the Inset sizes. Android 9 way does make the UI appear without pressing OK with first key press in Messaging and Contacts app, but OK doesn't work for selecting underlined text or selecting downlist items in Contacts. 

I wonder if Antdroid 9 + DPAD_CENTER is yet another option that may do the trick? 

The only mode that works acceptably well is DPAD_CENTER emulation.

**Comment by sspanak** [comment]
At last, some good news!

Selecting the suggestions wasn't working, because the code for the OK key became a mess. But it doesn't matter, I've cleaned it up and included it in the next release train. 

Here is [v30.30](https://github.com/sspanak/tt9/releases/download/v1.0-inital/tt9-v30.30.apk). It should work out-of-the-box. If it doesn't, just make sure the "Always on Top" option in Settings -> Appearance (at the bottom) is on. This is in case the XP3900 wasn't detected properly.

As a bonus, you get an early preview of the new stuff that is coming out soon.

**Comment by neutralinsomniac** [comment]
Works flawlessly over here. Thank you so much for your dedication and attention to detail with this keyboard. It makes using my flip phone an absolute dream.

**Comment by asherGe** [comment]
Fantastico! Great work. Congrats on the breakthrough and thanks for your perseverance!  Sonim should make tt9 the default keyboard and compensate you with hefty sum!

**Comment by sspanak** [comment]
> Works flawlessly over here. Thank you so much for your dedication and attention to detail with this keyboard. It makes using my flip phone an absolute dream.

> Fantastico! Great work. Congrats on the breakthrough and thanks for your perseverance!

I couldn't have done it myself. I must thank you for cooperating, it was truly enjoyable teamwork!

>  Sonim should make tt9 the default keyboard and compensate you with hefty sum!

I wouldn't oppose that. 😛 And, I am open to collaboration with anyone. My philosophy is everyone is welcome to help in makeing Traditional T9 the best keyboard.

**Comment by wazroth** [comment]
So far 30.30 is working a treat for me on my xp3900. Thanks for the hard work! Will update if I notice anything funny about OK behavior.

**Comment by sspanak** [comment]
With the new version just around the corner, I think it is time to finally close this issue. I believe all discussed problems are solved.

Again, thank you all for testing and for throwing ideas around, eventually leading to fixing all bugs. I couldn't have done it on my own.
